libs/api/ebpf_api.cpp

```diff
@@ -176,47 +179,59 @@ resolve_maps_in_byte_code(ebpf_handle_t program_handle, ebpf_code_buffer_t& byte
 
         uint64_t imm =
             static_cast<uint64_t>(first_instruction.imm) | (static_cast<uint64_t>(second_instruction.imm) << 32);
-        instruction_offsets.push_back(index - 1);
-        map_handles.push_back(imm);
+
+        // Collect set of instructions to patch with the value to replace.
+        instruction_offsets_to_map_handles[index - 1] = imm;
+
+        // Collect set if map handles.
+        map_handles_to_map_addresses[imm] = 0;
     }
 
-    if (map_handles.size() == 0) {
+    if (instruction_offsets_to_map_handles.size() == 0) {
         return ERROR_SUCCESS;
     }
 
     ebpf_protocol_buffer_t request_buffer(
-        offsetof(ebpf_operation_resolve_map_request_t, map_handle) + sizeof(uint64_t) * map_handles.size());
+        offsetof(ebpf_operation_resolve_map_request_t, map_handle) +
+        sizeof(uint64_t) * map_handles_to_map_addresses.size());
 
     ebpf_protocol_buffer_t reply_buffer(
-        offsetof(ebpf_operation_resolve_map_reply_t, address) + sizeof(uint64_t) * map_handles.size());
+        offsetof(ebpf_operation_resolve_map_reply_t, address) + sizeof(uint64_t) * map_handles_to_map_addresses.size());
 
     auto request = reinterpret_cast<ebpf_operation_resolve_map_request_t*>(request_buffer.data());
     auto reply = reinterpret_cast<ebpf_operation_resolve_map_reply_t*>(reply_buffer.data());
     request->header.id = ebpf_operation_id_t::EBPF_OPERATION_RESOLVE_MAP;
     request->header.length = static_cast<uint16_t>(request_buffer.size());
     request->program_handle = reinterpret_cast<uint64_t>(program_handle);
 
-    for (size_t index = 0; index < map_handles.size(); index++) {
-        if (map_handles[index] > get_map_descriptor_size()) {
+    size_t index = 0;
+    for (auto& mapping : map_handles_to_map_addresses) {
+
+        if (mapping.first > get_map_descriptor_size()) {
             return ERROR_INVALID_PARAMETER;
         }
-        request->map_handle[index] = get_map_handle_at_index((int)map_handles[index] - 1);
+        request->map_handle[index++] = get_map_handle_at_index((int)mapping.first - 1);
     }
 
     uint32_t result = invoke_ioctl(device_handle, request_buffer, reply_buffer);
     if (result != ERROR_SUCCESS) {
         return result;
     }
 
-    for (size_t index = 0; index < map_handles.size(); index++) {
-        ebpf_inst& first_instruction = instructions[instruction_offsets[index]];
-        ebpf_inst& second_instruction = instructions[instruction_offsets[index] + 1];
+    index = 0;
+    for (auto& mapping : map_handles_to_map_addresses) {
```

Comment:
rather than using .first and .second, it's much more readable to use structured bindings (see https://stackoverflow.com/questions/14555751/how-to-iterate-over-a-c-stl-map-data-structure-using-the-auto-keyword).
```suggestion
    for (auto& [map_handle, map_address] : map_handles_to_map_addresses) {
```